glib/src/value.rs

```diff
@@ -56,31 +51,223 @@ use crate::gstring::GString;
     }
 }
 
-impl error::Error for GetError {}
+impl error::Error for ValueTypeMismatchError {}
+
+/// Generic `Value` type checker for types.
+pub struct GenericValueTypeChecker<T>(std::marker::PhantomData<T>);
+
+unsafe impl<T: StaticType> ValueTypeChecker for GenericValueTypeChecker<T> {
+    type Error = ValueTypeMismatchError;
+
+    fn check(value: &Value) -> Result<(), Self::Error> {
+        unsafe {
+            if gobject_ffi::g_type_check_value_holds(&value.0, T::static_type().to_glib())
+                == ffi::GFALSE
+            {
+                Err(ValueTypeMismatchError::new(
+                    Type::from_glib(value.0.g_type),
+                    T::static_type(),
+                ))
+            } else {
+                Ok(())
+            }
+        }
+    }
+}
+
+/// An error returned from the [`get`](struct.Value.html#method.get)
+/// function on a [`Value`](struct.Value.html) for optional types.
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum ValueTypeMismatchOrNoneError {
+    WrongValueType(ValueTypeMismatchError),
+    UnexpectedNone,
+}
+
+impl fmt::Display for ValueTypeMismatchOrNoneError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            ValueTypeMismatchOrNoneError::WrongValueType(err) => err.fmt(f),
+            ValueTypeMismatchOrNoneError::UnexpectedNone => write!(f, "Unexpected None",),
+        }
+    }
+}
+
+impl error::Error for ValueTypeMismatchOrNoneError {}
+
+impl From<ValueTypeMismatchError> for ValueTypeMismatchOrNoneError {
+    fn from(err: ValueTypeMismatchError) -> Self {
+        ValueTypeMismatchOrNoneError::WrongValueType(err)
+    }
+}
+
+/// Generic `Value` type checker for optional types.
+pub struct GenericValueTypeOrNoneChecker<T>(std::marker::PhantomData<T>);
+
+unsafe impl<T: StaticType> ValueTypeChecker for GenericValueTypeOrNoneChecker<T> {
+    type Error = ValueTypeMismatchOrNoneError;
+
+    fn check(value: &Value) -> Result<(), Self::Error> {
+        GenericValueTypeChecker::<T>::check(value)?;
+
+        unsafe {
+            // Values are always zero-initialized so even if pointers are only 32 bits then the
+            // whole 64 bit value will be 0 for NULL pointers.
+            if value.0.data[0].v_uint64 == 0 {
+                return Err(ValueTypeMismatchOrNoneError::UnexpectedNone);
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// Trait to retrieve the contained value from a `Value`.
+///
+/// Usually this would not be used directly but from the [`get`](struct.Value.html#method.get)
+/// function on a [`Value`](struct.Value.html)
+pub unsafe trait FromValue<'a>: Sized {
+    /// Value type checker.
+    type Checker: ValueTypeChecker<Error = Self::Error>;
+    /// Error returned from `Checker`.
+    // FIXME: This has to be a separate associated type because we can't do
+    // `where T: FromValue<Checker::Error = ValueTypeMismatchOrNoneError>`
+    type Error: std::error::Error + Send + Sized + 'static;
```

Comment:
This sucks a bit. @fengalin do you know a solution so the error type doesn't have to be effectively duplicated here?

Reply:
I think you can define a 2 steps constraint, see lines 32 & 33 in https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9808b1508c31f01c1f525c9ce3a27908

Reply:
That works but you can't do

```rust
pub trait ValueTypeOptional:
    ValueType + ToValueOptional + FromValue<'static, Checker::Error = ValueTypeMismatchOrNoneError> + StaticType
{
}

unsafe impl<'a, T: FromValue<'a, Checker::Error = ValueTypeMismatchOrNoneError> + StaticType> FromValue<'a>
    for Option<T>
{
}
```

The `::` give parser errors in both positions there

Reply:
My solution addresses the 2d case AFAICT.

Reply:
Not really. Any `<` in those positions causes a parser error too :)

For example

```rust
unsafe impl<'a, T: FromValue<'a, <Checker as ValueTypeChecker>::Error = ValueTypeMismatchOrNoneError> + StaticType> FromValue<'a>
    for Option<T>
{
}
```

Gives

```
error: expected one of `(`, `,`, `::`, `:`, `<`, or `>`, found `=`
   --> glib/src/value.rs:211:71
    |
211 | unsafe impl<'a, T: FromValue<'a, <Checker as ValueTypeChecker>::Error = ValueTypeMismatchOrNoneError> + StaticType> FromValue<'a>
    |                                                                 ----- ^ expected one of `(`, `,`, `::`, `:`, `<`, or `>`
    |                                                                 |
    |                                                                 maybe try to close unmatched angle bracket
```

I'm not sure if this is intentional. In `where` clauses this works btw, so that would be an option for the `impl` blocks but doesn't work for super-traits.

Reply:
>  but doesn't work for super-traits.
where clauses for super-traits are not stable, that's why.

Reply:
That's why I used:

```rust
impl<T, C> FromValue for Option<T>
where
    T: FromValue<Checker = C>,
    C: TypeChecker<Error = GenericError>,
{
    type Checker = GenericChecker;
}
```

With your example, I guess, it would become:

```rust
unsafe impl<'a, T, C> FromValue<'a> for Option<T>
where
    T: FromValue<'a, Checker = C> + StaticType,
    C: ValueTypeChecker<Error = ValueTypeMismatchOrNoneError>,
{
}
```

But that doesn't address your first use case, maybe a solution would to use additional traits and blanket implementations like in https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=0d01fe15092b6e7557a1c0aa4e645712

I have to go for an hour or so.

Reply:
Hmm... yes that could potentially work. I'll give it a try.

Reply:
Nope, doesn't work. Your version doesn't work because `C` is not part of the trait or implementor type. And the following doesn't work because the constraint is in a `where` clause and not in the `impl` bounds:

```rust
pub trait OptionalValueTypeChecker: ValueTypeChecker {}
impl<T: ValueTypeChecker<Error = ValueTypeMismatchOrNoneError>> OptionalValueTypeChecker for T {}

/// Blanket implementation for all optional types.
unsafe impl<'a, T: FromValue<'a> + StaticType> FromValue<'a>
    for Option<T>
    where <T as FromValue<'a>>::Checker: OptionalValueTypeChecker,
{
    type Checker = GenericValueTypeChecker<T>;
    type Error = ValueTypeMismatchError;

    unsafe fn from_value(value: &'a Value) -> Self {
        if let Err(ValueTypeMismatchOrNoneError::UnexpectedNone) = T::Checker::check(value) {
            None
        } else {
            Some(T::from_value(value))
        }
    }
}
```

Gives

```
222 |         if let Err(ValueTypeMismatchOrNoneError::UnexpectedNone) = T::Checker::check(value) {
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^    ------------------------ this expression has type `std::result::Result<(), <T as FromValue<'a>>::Error>`
    |                    |
    |                    expected associated type, found enum `ValueTypeMismatchOrNoneError`
    |
    = note: expected associated type `<T as FromValue<'a>>::Error`
                          found enum `ValueTypeMismatchOrNoneError`
```

I guess let's keep it like this for now and we can figure out a nicer way after it's merged.

Reply:
> Your version doesn't work because C is not part of the trait or implementor type.

Maybe the use case is different, but it runs in my playground. `C` can be used thanks to `impl<'a, T, C>`.
I wanted to try it on your branch, but it seems you found a solution.

Reply:
Something to look into at a later time :) Maybe you can take a look after it's merged? Worst case we can remove some more code from all the crates afterwards.